crates/table/src/read_column.rs

```diff
@@ -339,6 +341,30 @@ impl_read_column_via_from! {
     i256 => Box<i256>;
 }
 
+/// SAFETY: `is_compatible_type` only returns true for sum types,
+/// and any sum value stores the tag first in BFLATN.
+unsafe impl ReadColumn for SumTag {
+    fn is_compatible_type(ty: &AlgebraicTypeLayout) -> bool {
+        matches!(ty, AlgebraicTypeLayout::Sum(_))
+    }
+
+    unsafe fn unchecked_read_column(row_ref: RowRef<'_>, layout: &ProductTypeElementLayout) -> Self {
+        debug_assert!(Self::is_compatible_type(&layout.ty));
+
+        let (page, offset) = row_ref.page_and_offset();
+        let col_offset = offset + PageOffset(layout.offset);
```

Comment:
If we're going to write this method in the maximally-defensive style (using `mem::size_of::<Self>()` and reading an array, rather than just reading a `u8`), we might as well also respect `SumTypeLayout::offset_of_tag` here.